### PR TITLE
Write project file

### PR DESCRIPTION
**Pull request by morettid**
This pull request supersedes pull request #76.

TODO specified in commit 0369c6809f563596b4f78f37dc4cd129537f8eb9

### CODE REVIEW

File: stem/IO/kratos_model_io.py [hunk 1]
```diff
@@ -0,0 +1,547 @@
...
+        """
+        Writes nodes to mdpa format.
+
+        Args:
+            - model (:class:`stem.model.Model`): the model object containing the info on the nodes.
+
+        Returns:
+            - block_text (List[str]): list of strings for the mdpa file. Each element is a line in the mdpa file.
+        """
+        nodes_dict = model.get_all_nodes()
```
Comment (aronnoordam) [inline stem/IO/kratos_model_io.py:442]: get_all_nodes is fine for now, but I want to change this in the future, to the same structure as Geometry.points

File: stem/IO/kratos_model_io.py [hunk 2]
```diff
@@ -0,0 +1,547 @@
...
+
+        Returns:
+            - block_text (List[str]): list of strings for the mdpa file. Each element is a line in the mdpa file.
+        """
+        # get the unique ids in material and conditions
+        ids_to_write = list(set([mp.id for mp in model.get_all_model_parts()]))
+        # get the unique ids and write properties
+
+        block_text = []
+        for _id in np.sort(ids_to_write):
```
Comment (aronnoordam) [inline stem/IO/kratos_model_io.py:470]: will the properties also be written in the same sorted order as the ids?

File: stem/model.py [hunk 3]
```diff
@@ -295,6 +295,16 @@ def synchronise_geometry(self):
         # get the complete geometry
         self.__get_geometry_from_geo_data(self.gmsh_io.geo_data)
 
+    def set_mesh_size(self, element_size:float):
+        """Set the element size to dimension (m).
```
Comment (aronnoordam) [inline stem/model.py:299]: can you format the docstring correctly in this whole file? Like how you have in kratos_model_io.py

File: stem/utils.py [hunk 4]
```diff
@@ -112,3 +112,35 @@ def is_point_between_points(point:Sequence[float], start_point:Sequence[float],
 
         # Check if the scalar projection is between 0 and 1 (inclusive)
         return 0 <= scalar_projection <= 1
+
+    @staticmethod
+    def merge(a: Dict[str, Any], b: Dict[str, Any], path: Union[List[str], Any] = None):
```
Comment (aronnoordam) [inline stem/utils.py:117]: please reformat docstring and add tests for merge

File: tests/test_kratos_io.py [hunk 5]
```diff
@@ -0,0 +1,206 @@
+import json
+from typing import List
+
+import numpy.testing as npt
+from stem.load import LineLoad
+from stem.model import Model
+from stem.model_part import *
+from stem.output import NodalOutput, GaussPointOutput, GiDOutputParameters, Output
+from stem.soil_material import OnePhaseSoil, LinearElasticSoil, SaturatedBelowPhreaticLevelLaw
+from gmsh_utils import gmsh_IO
```
Comment (aronnoordam) [inline tests/test_kratos_io.py:12]: please reorder inports: python build ins -> external packages -> project imports

File: tests/test_kratos_io.py [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+        # set up solver settings
+        analysis_type = AnalysisType.MECHANICAL_GROUNDWATER_FLOW
+        solution_type = SolutionType.DYNAMIC
+        time_integration = TimeIntegration(start_time=0.0, end_time=1.0, delta_time=0.1, reduction_factor=0.5,
+                                           increase_factor=2.0, max_delta_time_factor=500)
+        convergence_criterion = DisplacementConvergenceCriteria(displacement_relative_tolerance=1e-5,
+                                                                displacement_absolute_tolerance=1e-7)
+        strategy_type = NewtonRaphsonStrategy(min_iterations=5, max_iterations=30, number_cycles=50)
+        scheme_type = NewmarkScheme(newmark_beta=0.35, newmark_gamma=0.4, newmark_theta=0.6)
+        linear_solver_settings = Amgcl(tolerance=1e-8, max_iterations=500, scaling=True)
```
Comment (aronnoordam) [inline tests/test_kratos_io.py]:
```suggestion
        linear_solver_settings = Amgcl(tolerance=1e-8, max_iterations=500, scaling=False)
```

File: tests/test_kratos_io.py [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+                soil layer and a line load.
+        """
+        model = create_default_2d_model_and_mesh
+        kratos_io = KratosIO(ndim=model.ndim)
+
+        actual_dict = kratos_io.write_material_parameters_json(model=model, output_folder="dir_test")
+        expected_dict = json.load(open("tests/test_data/expected_MaterialParameters.json", 'r'))
+        TestUtils.assert_dictionary_almost_equal(expected_dict, actual_dict)
+
+    def test_write_mdpa_file(
```
Comment (aronnoordam) [inline tests/test_kratos_io.py:190]: can you also add a test for 3D

Comment (aronnoordam) [reply tests/test_kratos_io.py:190]: only for the mpda, becase the material parameters and project parameters are already tested in other io's I think

File: tests/test_kratos_model_io.py [hunk 8]
```diff
@@ -0,0 +1,128 @@
+import json
+import numpy.testing as npt
+from stem.load import LineLoad
+from stem.model import Model
+from stem.model_part import *
+from stem.soil_material import OnePhaseSoil, LinearElasticSoil, SaturatedBelowPhreaticLevelLaw
+from gmsh_utils import gmsh_IO
+
+import pytest
```
Comment (aronnoordam) [inline tests/test_kratos_model_io.py:11]: please reorder imports -> build-ins -> external packages -> project imports

File: run_stem/run_kratos_model.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
+import os
+import sys
+
+# IMPORTANT !!!! SPECIFY LOCAL PATH TO KRATOS...
+pth_kratos = r"C:\Users\morettid\OneDrive - TNO\Desktop\projects\STEM"
```
Comment (aronnoordam) [inline run_stem/run_kratos_model.py:5]:
```suggestion
pth_kratos = r"C:\Please\Define\Installation\Directory\Of\Kratos"
```

File: run_stem/run_kratos_model.py [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+sys.path.append(os.path.join(pth_kratos, r"KratosGeoMechanics\libs"))
+
+import KratosMultiphysics.GeoMechanicsApplication
+from KratosMultiphysics.GeoMechanicsApplication.geomechanics_analysis import (
+    GeoMechanicsAnalysis,
+)
+
+project_folder = "dir_test"
+
+if run_only:
```
Comment (aronnoordam) [inline run_stem/run_kratos_model.py:22]: I think this bool can removed, as you don't do anything else

File: run_stem/run_kratos_model.py
Comment (aronnoordam) [inline run_stem/run_kratos_model.py:1]: can you rename to demo_run_kratos_model.py

File: tests/test_kratos_io.py [hunk 11]
```diff
@@ -0,0 +1,290 @@
...
+        model = create_default_3d_model_and_mesh
+        kratos_io = KratosIO(ndim=model.ndim)
+        model.project_parameters = create_default_solver_settings
+
+        actual_text = kratos_io.write_mesh_to_mdpa(
+            model=model,
+            mesh_file_name="test_mdpa_file_3d.mdpa",
+            output_folder="dir_test"
+        )
+        if sys.platform == "win32":
```
[the remark this answers is not in the document]
Comment (aronnoordam) [reply tests/test_kratos_io.py:282]: for example @pytest.mark.skipif(not SHOW_PLOTS, reason="Plotting is disabled")

File: tests/test_utils.py [hunk 12]
```diff
@@ -79,5 +83,79 @@ def test_is_in_between_3d(self):
...
+        # new keys are added, common keys are expanded
+        seq1 = [1, 2, 3, 4]
+        seq2 = (100, 200, 300)
+
+        actual_sequence = list(Utils.chain_sequence([seq1, seq2]))
+        expected_sequence = [1, 2, 3, 4, 100, 200, 300]
+
+        npt.assert_equal(expected_sequence, actual_sequence)
+
+    def test_merge(self):
```
Comment (aronnoordam) [inline tests/test_utils.py:119]: raise error when any of two values with same key are non- sequences

File: tests/test_kratos_io.py [hunk 13]
```diff
@@ -17,6 +17,8 @@
     NewtonRaphsonStrategy, NewmarkScheme, Amgcl, StressInitialisationType, SolverSettings, Problem
 from tests.utils import TestUtils
 
+IS_WINDOWS = sys.platform == "win32"
```
Comment (aronnoordam) [inline tests/test_kratos_io.py:20]: can you switch to IS_UBUNTU? Because both windoes 32 and 64 should be supported

**Comment by morettid** [comment]
@aronnoordam

Done!

**Review by aronnoordam — APPROVED** [review]

**Pull request merged**